lib/_stream_pair.js

```diff
@@ -38,12 +38,13 @@ class DuplexSocket extends Duplex {
   }
 }
 
-function makeDuplexPair() {
-  const clientSide = new DuplexSocket();
-  const serverSide = new DuplexSocket();
+function DuplexPair() {
+  if (!(this instanceof DuplexPair))
```

Comment:
Can we avoid adding this for new APIs? We may as well use a `class`, which would force `new` anyway.

Reply:
Yeah I'll check it out.
Does Node.js do it this way anywhere else? I was trying too copy how the other stream classes work.

Reply:
We use `class` for almost everything since `class` has been introduced into the language. Unfortunately, we cannot convert this old-style pattern to ES6 classes without a breaking change.

Reply:
@addaleax Would it be possible to get a deprecation notice for that, then? Or is there some documentation on what causes the breaking change? (My understanding was that `class` is just syntactic sugar.)

Reply:
> @addaleax Would it be possible to get a deprecation notice for that, then?

I don’t think this is something that we’re ever realistically going to be able to change, so there’s no real reason to deprecated it, I assume.

> Or is there some documentation on what causes the breaking change? (My understanding was that `class` is just syntactic sugar.)

It’s the fact that classes *have* to be invoked with `new`, and will throw otherwise. That”s not the case in the current code here.

Reply:
@addaleax Thanks! I'm working on that now.

Sort of off-topic follow-up, is calling builtins without `new` discouraged too then? e.g. `const now = Date();`? I've been in discussions with a couple core members on the IRC channel, whos said more-or-less "new is obsolete these days" but I assume that was just a personal opinion around various programming styles.

Reply:
@awwright `Date` is special, because `Date()` and `new Date()` do different things :)

But yes, for a function that implements a pattern like this (returns an instance of itself even when called without `new`), I would recommend using `new` because that allows transitioning to classes.

Reply:
I'm trying this code:
```javascript
class DuplexPair {
  [Symbol.iterator] = Array.prototype[Symbol.iterator];
  length = 2;
  constructor() {
    const clientSide = this[0] = new DuplexSide();
    const serverSide = this[1] = new DuplexSide();
    this.serverSide = clientSide[kOtherSide] = serverSide;
    this.clientSide = serverSide[kOtherSide] = clientSide;
  }
}
```

This passes the `test-release` suite, but the acorn dependency can't grok it ("SyntaxError: unexpected token (42:20)" which is at the first closing square bracket). Am I missing something here?

Reply:
@awwright Yeah, we’ve run into trouble with acorn in the past for private properties as well, so I guess this falls into a similar category :/ Fwiw, I think if you do this, you could also use `class DuplexPair extends Array`, to automatically get `length` and the iterator property…